### PR TITLE
zed.d/pool_import-led.sh: fix for current zpool scripts (and a few bonus touchups)

### PR DESCRIPTION
**Pull request by nabijaczleweli**
### Motivation and Context
#11934 and a few others I noticed

### Description
I mean, it's what it says on the tins.

### How Has This Been Tested?
Looked at it, and tested on fake data, but I don't have a system with SES enclosures, so it'd be great if someone who does could run statechange-led.sh on pool import :)

The rest should be relative non-brainers.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes. – let's hope none break
- [ ] I have run the ZFS Test Suite with this change applied. – CI take my hand
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [inline cmd/zed/zed.d/statechange-led.sh:100]: i blindly translated this, but it could be better served as ONLINE => 0, anything else => 1 maybe

File: cmd/zed/zed.d/data-notify.sh
Comment (mskarbek) [inline cmd/zed/zed.d/data-notify.sh:28]: `mktemp zed.XXXXXXXXXX`?

Comment (nabijaczleweli) [reply cmd/zed/zed.d/data-notify.sh:28]: I don't really see the point in deviating from the default for a file that's `rm -f`ed like 15 lines below in all cases (also, it'd be `mktemp -t zed.XXXXXXXXXX`)

File: cmd/zed/zed.d/statechange-led.sh
Comment (behlendorf) [reply cmd/zed/zed.d/statechange-led.sh:100]: There are a couple valid states not listed here, (e.g. REMOVED, SPLIT, OFFLINE) so I'm not sure "anything else => 1`" would be quite right either.  A third "we don't care" state might actually be preferable.  But I agree, let's leave that for another time and stick with the direct translation.

File: cmd/zed/zed.d/data-notify.sh
Comment (behlendorf) [reply cmd/zed/zed.d/data-notify.sh:28]: The only advantage I can see is it makes it easy to identify these zed file if they somehow fail to get cleaned up.  The existing manual scheme seems fine, but I don't have any objection to just using `mktemp`.

**Review by behlendorf — COMMENTED** [review]
I'm fairly certain this script was working but let's definitely make sure.  cc: @tonyhutter.

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh:100]: In that case, may it be preferable to explicitly ignore `[ -z "$val" ]`, instead of the current behaviour which would to write a null string to the `/fault` file five times?

**Comment by nabijaczleweli** [comment]
Added fix for #11954, too

File: cmd/zed/zed.d/zed-functions.sh
Comment (behlendorf) [inline cmd/zed/zed.d/zed-functions.sh:514]: Dropping the `printf` would mean this function no longer accepts the GUID as either decimal or hex.  We should keep that functionality even though the only callers we provide all pass `$ZEVENT_POOL_GUID`.  Others may have written their own scripts which depend on this behavior.

Comment (nabijaczleweli) [reply cmd/zed/zed.d/zed-functions.sh]: urgh, this worked in my tests because GAWK supports hex integers as an extension, but MAWK (and, presumably, POSIX AWK) doesn't seem to

updated to retain the printf

**Review by behlendorf — APPROVED** [review]

**Comment by tonyhutter** [comment]
I'll take a look at this

**Comment by ZNikke** [comment]
Ah, my travel down report-issue-lane started with looking at statechange-led and wondering how/why the process was supposed to work.

I have a patch that generalizes the led-handling so statechange-led can light the locate led for OFFLINE devices as well, but I'll hold off on that until this gets landed.

Let me know if this needs more testers, as we have enclosures with blinkenlights...

**Comment by tonyhutter** [comment]
While reviewing this patch, I realized that the code that code causing the bug (#11934) will never actually get called, so I ended up opening  a PR (https://github.com/openzfs/zfs/pull/11964) to remove that code.  My vote would be to use #11964 as the fix, close this PR, and move your "few bonus touchups" to a separate PR.

**Comment by tonyhutter** [comment]
Spoke too soon - as @nabijaczleweli points out in #11964, removing the `process_pool()` in isn't going to work for `pool_import-led.sh` case.  I'm looking into a different fix...

**Comment by nabijaczleweli** [comment]
What other fix? I already wrote one, it's right here.

**Comment by tonyhutter** [comment]
@nabijaczleweli this simplified patch uses `zpool status -c upath,fault_led` and works on my enclosure.  Does it work for you?
```patch
diff --git a/cmd/zed/zed.d/statechange-led.sh b/cmd/zed/zed.d/statechange-led.sh
index e656e125d..4b25a55f9 100755
--- a/cmd/zed/zed.d/statechange-led.sh
+++ b/cmd/zed/zed.d/statechange-led.sh
@@ -115,17 +115,24 @@ process_pool()
        pool="$1"
        rc=0
 
-       # Lookup all the current LED values and paths in parallel
        #shellcheck disable=SC2016
-       cmd='echo led_token=$(cat "$VDEV_ENC_SYSFS_PATH/fault"),"$VDEV_ENC_SYSFS_PATH",'
-       out=$($ZPOOL status -vc "$cmd" "$pool" | grep 'led_token=')
+
+       # The output will be the vdevs only (from "grep '/dev/'"):
+       #
+       #    U45     ONLINE       0     0     0   /dev/sdk          0
+       #    U46     ONLINE       0     0     0   /dev/sdm          0
+       #    U47     ONLINE       0     0     0   /dev/sdn          0
+       #    U50     ONLINE       0     0     0  /dev/sdbn          0
+       #
+       out="$(ZPOOL_SCRIPTS_AS_ROOT=1 $ZPOOL status -c upath,fault_led $pool | grep '/dev/')"
 
        #shellcheck disable=SC2034
        echo "$out" | while read -r vdev state read write chksum therest; do
                # Read out current LED value and path
-               tmp=$(echo "$therest" | sed 's/^.*led_token=//g')
-               vdev_enc_sysfs_path=$(echo "$tmp" | awk -F ',' '{print $2}')
-               current_val=$(echo "$tmp" | awk -F ',' '{print $1}')
+               # Get dev name (like 'sda')
+               dev=$(basename $(echo "$therest" |  awk '{print $(NF-1)}'))
+               vdev_enc_sysfs_path=$(realpath /sys/class/block/$dev/device/enclosure_device*)
+               current_val=$(echo "$therest" | awk '{print $NF}')
 
                if [ "$current_val" != "0" ] ; then
                        current_val=1
@@ -171,7 +178,7 @@ if [ -n "$ZEVENT_VDEV_ENC_SYSFS_PATH" ] && [ -n "$ZEVENT_VDEV_STATE_STR" ] ; the
        vdev=$(basename "$ZEVENT_VDEV_PATH")
        check_and_set_led "$ZEVENT_VDEV_ENC_SYSFS_PATH/fault" "$val"
 else
-       # Process the entire pool
+       # Process the entire pool (needed for pool_import-led.sh)
        poolname=$(zed_guid_to_pool "$ZEVENT_POOL_GUID")
        process_pool "$poolname"
 fi
```

**Comment by nabijaczleweli** [comment]
Hm, yeah, that is a better approach. Had to fix dm_get_underlying_path() in the mean-time, but.

I've updated the PR, it'd be great if you could try it again.

File: cmd/zed/zed.d/zed-functions.sh
Comment (tonyhutter) [reply cmd/zed/zed.d/zed-functions.sh:514]: > updated to retain the printf

I don't see the printf

Comment (nabijaczleweli) [reply cmd/zed/zed.d/zed-functions.sh]: odd, coulda swore i added it back when i made the comment

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:144]: This section assumes a fixed ordering of the fields, but that's not the case.  A bad disk will have some kind of message after the checksum field, like:
```
$ ZPOOL_SCRIPTS_AS_ROOT=1 ZPOOL status -c "upath,fault_led"
...
	    U83     FAULTED      0     0     0  external device fault  /dev/sdcj          2
```

Also, this is going to an issue:
```
 			sysfs_path="$(readlink -f "/sys/block/$dev/device/enclosure_device")" 
```
The enclosure directory name is `enclosure_device${vendor_stuff}`, like `/sys/class/block/sdcj/device/enclosure_device:SLOT 84 6E  /` (that is not a typo - there are two spaces after the "6E" in the directory name).  I've seen others that use different format, but they all begin with `enclosure_device`.

Would you mind using the patch I posted instead?  It will deal with these issues, and it correctly worked for me on the enclosure I tested on.

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:3]: The grammar was correct before.  VDEV is an acronym for "Virtual Device" or "Virtual Disk" and acronyms are typically capitalized.  "VDEV's" was also correct since it's the possessive form of the noun.

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:6]: What is the point of this change?

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:29]: What is this for?

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:71]: What is the point of this change?

Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh:71]: As noted: seq is a very spotty GNU extension (and also not checked for at the top), so when not available this resolves to `for _ in ; do` which I assume you can guess why is less than optimal. (Plus, it's literally five numbers, do you really need to spawn a process for it?)

File: cmd/zed/zed.d/statechange-led.sh
Comment (tonyhutter) [inline cmd/zed/zed.d/statechange-led.sh:153]: This check is a good idea, but would you mind moving it into `check_and_set_led()`?  That way it could catch an empty `$val` for both the single VDEV and `process_pool()` cases.

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh:3]: This script toggles enclosure fault LEDs on multiple vdevs at the point when a particular vdev's pool experiences a state change, or "toggles vdevs' enclosure fault LEDs when their pool's state changes"

vdev is not an acronym (that'd be "VD"), but, rather, a shorthand-turned-object-class for "virtual device" as ZFS implements it, consult zpoolconcepts(8), or zpoolprops(8), or zpool-status(8), or

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh:6]: This manipulates many fault LEDs at a time, also making terminology consistent with the rest

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh]: nothing anymore

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh]: ugh, this is much worse than I thought it would be. Used your patch instead of this

File: cmd/zed/zed.d/statechange-led.sh
Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh]: Done

File: cmd/zed/zed.d/statechange-led.sh
Comment (gdevenyi) [reply cmd/zed/zed.d/statechange-led.sh:73]: Has checkbashisms said that?

Comment (nabijaczleweli) [reply cmd/zed/zed.d/statechange-led.sh:73]: Hm, I've misremembered apparently, according to FreeBSD's seq(1) it's had a tumultuous history (as ported from GNU shellutils and/or P9) and should be available on most recent fully-configured UNIXes, but I've also definitely used systems without it in recent years.
Plus, if the portability to small systems doesn't appeal to you, then the performance gains of not using a process to generate 5 numbers that are, in sum, shorter than the process substitution definitely will.

and dunno, haven't run checkbashisms

**Review by tonyhutter — APPROVED** [review]
I re-tested statechange-led.sh with your latest push and works as expected.  Thanks for sticking with this! :+1:

**Pull request closed without merging**
